src/guest_memory.rs

```diff
@@ -175,19 +181,89 @@ pub trait GuestMemory {
     fn num_regions(&self) -> usize;
 
     /// Return the region containing the specified address or None.
-    fn find_region(&self, GuestAddress) -> Option<&Self::R>;
+    fn find_region(&self, addr: GuestAddress) -> Option<&Self::R>;
 
-    /// Perform the specified action on each region.
+    /// Perform the specified action on each region mutably.
     /// It only walks children of current region and do not step into sub regions.
-    fn with_regions<F>(&self, cb: F) -> Result<()>
+    fn with_regions<F, E>(&self, cb: F) -> std::result::Result<(), E>
     where
-        F: Fn(usize, &Self::R) -> Result<()>;
+        F: Fn(usize, &Self::R) -> std::result::Result<(), E>;
 
     /// Perform the specified action on each region mutably.
     /// It only walks children of current region and do not step into sub regions.
-    fn with_regions_mut<F>(&self, cb: F) -> Result<()>
+    fn with_regions_mut<F, E>(&self, cb: F) -> std::result::Result<(), E>
     where
-        F: FnMut(usize, &Self::R) -> Result<()>;
+        F: FnMut(usize, &Self::R) -> std::result::Result<(), E>;
+
+    /// Applies two functions, specified as callbacks, on the inner memory regions.
+    ///
+    /// # Arguments
+    /// * `init` - Starting value of the accumulator for the `foldf` function.
+    /// * `mapf` - "Map" function, applied to all the inner memory regions. It returns an array of
+    ///            the same size as the memory regions array, containing the function's results
+    ///            for each region.
+    /// * `foldf` - "Fold" function, applied to the array returned by `mapf`. It acts as an
+    ///             operator, applying itself to the `init` value and to each subsequent elemnent
+    ///             in the array returned by `mapf`.
+    ///
+    /// # Examples
+    ///
+    /// * Compute the total size of all memory mappings in KB by iterating over the memory regions
+    ///   and dividing their sizes to 1024, then summing up the values in an accumulator.
+    ///
+    /// ```
+    /// # #[cfg(feature = "backend-mmap")]
+    /// # fn test_map_fold() -> Result<(), ()> {
+    /// # use vm_memory::{GuestAddress, GuestMemory, GuestMemoryRegion, mmap::GuestMemoryMmap};
+    ///     let start_addr1 = GuestAddress(0x0);
+    ///     let start_addr2 = GuestAddress(0x400);
+    ///     let mem = GuestMemoryMmap::new(&vec![(start_addr1, 1024), (start_addr2, 2048)]).unwrap();
+    ///     let total_size = mem.map_and_fold(
+    ///         0,
+    ///         |(_, region)| region.len() / 1024,
+    ///         |acc, size| acc + size
+    ///     );
+    ///     println!("Total memory size = {} KB", total_size);
+    ///     Ok(())
+    /// # }
+    /// ```
+    fn map_and_fold<F, G, T>(&self, init: T, mapf: F, foldf: G) -> T
+    where
+        F: Fn((usize, &Self::R)) -> T,
+        G: Fn(T, T) -> T;
+
+    /// Returns true if the given address is within the memory range available to the guest.
```

Comment:
Comment not really related to the function. Can also fix later (and was probably just copied like that anyway :D).